src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -460,54 +460,56 @@ void SystemDictionary::validate_protection_domain(InstanceKlass* klass,
                                                   TRAPS) {
   // Now we have to call back to java to check if the initating class has access
+  assert(class_loader() != NULL, "Should not call this");
+  assert(protection_domain() != NULL, "Should not call this");
+
   JavaValue result(T_VOID);
   LogTarget(Debug, protectiondomain) lt;
   if (lt.is_enabled()) {
     ResourceMark rm(THREAD);
     // Print out trace information
     LogStream ls(lt);
     ls.print_cr("Checking package access");
-    if (class_loader() != NULL) {
-      ls.print("class loader: ");
-      class_loader()->print_value_on(&ls);
-    } else {
-      ls.print_cr("class loader: NULL");
-    }
-    if (protection_domain() != NULL) {
-      ls.print(" protection domain: ");
-      protection_domain()->print_value_on(&ls);
-    } else {
-      ls.print_cr(" protection domain: NULL");
-    }
+    ls.print("class loader: ");
+    class_loader()->print_value_on(&ls);
+    ls.print(" protection domain: ");
+    protection_domain()->print_value_on(&ls);
     ls.print(" loading: "); klass->print_value_on(&ls);
     ls.cr();
   }
 
-  // This handle and the class_loader handle passed in keeps this class from
-  // being unloaded through several GC points.
-  // The class_loader handle passed in is the initiating loader.
-  Handle mirror(THREAD, klass->java_mirror());
+  // We only have to call checkPackageAccess if there's a security manager installed.
+  if (java_lang_System::has_security_manager()) {
 
-  InstanceKlass* system_loader = vmClasses::ClassLoader_klass();
-  JavaCalls::call_special(&result,
-                         class_loader,
-                         system_loader,
-                         vmSymbols::checkPackageAccess_name(),
-                         vmSymbols::class_protectiondomain_signature(),
-                         mirror,
-                         protection_domain,
-                         THREAD);
+    // This handle and the class_loader handle passed in keeps this class from
+    // being unloaded through several GC points.
+    // The class_loader handle passed in is the initiating loader.
+    Handle mirror(THREAD, klass->java_mirror());
 
-  if (HAS_PENDING_EXCEPTION) {
-    log_debug(protectiondomain)("DENIED !!!!!!!!!!!!!!!!!!!!!");
-  } else {
-   log_debug(protectiondomain)("granted");
-  }
+    InstanceKlass* system_loader = vmClasses::ClassLoader_klass();
+    JavaCalls::call_special(&result,
+                           class_loader,
+                           system_loader,
+                           vmSymbols::checkPackageAccess_name(),
+                           vmSymbols::class_protectiondomain_signature(),
+                           mirror,
+                           protection_domain,
+                           THREAD);
 
-  if (HAS_PENDING_EXCEPTION) return;
+    if (HAS_PENDING_EXCEPTION) {
+      log_debug(protectiondomain)("DENIED !!!!!!!!!!!!!!!!!!!!!");
+    } else {
+     log_debug(protectiondomain)("granted");
+    }
```

Comment:
Did you intend leaving this in?

Reply:
Yes, why not? It's very useful in logging.

Reply:
There's no context for this final bit out of output. It could appear many lines after the earlier logging, and unless there is some locking around all this, could be interleaved with different PD checks. It seems to me that all the logging could occur after the Java call and include the success/failure. It could also include whether the check trivially succeeded due to there being no SM.

Reply:
Ok, I see what you're suggesting.  Yes, I'll move all the logging after the check. I don't see any reason to have it if we're not going to make the upcall.  Also fixed the test.